File: apps/mobile/android/app/src/main/java/com/audixa/NativePlayerModule.kt
package com.audixa

import android.net.Uri
import android.os.Handler
import android.os.Looper
import androidx.media3.common.MediaItem
import androidx.media3.common.PlaybackException
import androidx.media3.common.Player
import androidx.media3.exoplayer.ExoPlayer
import androidx.media3.exoplayer.SeekParameters
import com.facebook.react.bridge.*
import com.facebook.react.modules.core.DeviceEventManagerModule

/**
 * Native Player Module for Android
 *
 * Provides high-performance media playback using ExoPlayer (Media3) with:
 * - Precise seeking (ms-level accuracy)
 * - A-B loop support
 * - Background audio playback
 * - Media session integration
 */
class NativePlayerModule(private val reactContext: ReactApplicationContext) :
    ReactContextBaseJavaModule(reactContext), LifecycleEventListener {

    private var player: ExoPlayer? = null
    private var handler: Handler? = null
    private var timeUpdateRunnable: Runnable? = null

    private var loopEnabled = false
    private var loopStartMs: Long = 0
    private var loopEndMs: Long = 0

    private var currentStatus = "idle"

    // Time update frequency (50ms = 20Hz)
    private val timeUpdateIntervalMs = 50L

    override fun getName(): String = "NativePlayerModule"

    override fun initialize() {
        super.initialize()
        reactContext.addLifecycleEventListener(this)
        handler = Handler(Looper.getMainLooper())
    }

    // MARK: - Player Management

    @ReactMethod
    fun load(uri: String, autoPlay: Boolean) {
        handler?.post {
            loadInternal(uri, autoPlay)
        }
    }

    private fun loadInternal(uri: String, autoPlay: Boolean) {
        // Clean up previous player
        cleanup()

        // Create ExoPlayer with precise seeking
        player = ExoPlayer.Builder(reactContext)
            .setSeekParameters(SeekParameters.EXACT)
            .build()
            .apply {
                // Set up event listener
                addListener(playerListener)

                // Create and set media item
                val mediaItem = MediaItem.fromUri(Uri.parse(uri))
                setMediaItem(mediaItem)

                // Prepare and optionally play
                prepare()
                playWhenReady = autoPlay
            }

        // Start time updates
        startTimeUpdates()

        updateStatus("loading")
    }

    private val playerListener = object : Player.Listener {
        override fun onPlaybackStateChanged(playbackState: Int) {
            when (playbackState) {
                Player.STATE_READY -> {
                    if (player?.isPlaying == true) {
                        updateStatus("playing")
                    } else {
                        updateStatus("paused")
                    }
                }
                Player.STATE_ENDED -> {
                    if (loopEnabled) {
                        seekToInternal(loopStartMs)
                        player?.play()
                    } else {
                        updateStatus("ended")
                        sendEvent("onEnded", Arguments.createMap())
                    }
                }
                Player.STATE_BUFFERING -> {
                    sendEvent("onBuffering", Arguments.createMap().apply {
                        putBoolean("isBuffering", true)
                    })
                }
                Player.STATE_IDLE -> {}
            }
        }

        override fun onIsPlayingChanged(isPlaying: Boolean) {
            if (player?.playbackState == Player.STATE_READY) {
                updateStatus(if (isPlaying) "playing" else "paused")
            }
        }

        override fun onPlayerError(error: PlaybackException) {
            sendError(
                code = error.errorCode.toString(),
                message = error.message ?: "Playback error"
            )
        }
    }

    private fun startTimeUpdates() {
        timeUpdateRunnable = object : Runnable {
            override fun run() {
                sendTimeUpdate()
                handler?.postDelayed(this, timeUpdateIntervalMs)
            }
        }
        handler?.post(timeUpdateRunnable!!)
    }

    private fun stopTimeUpdates() {
        timeUpdateRunnable?.let { handler?.removeCallbacks(it) }
        timeUpdateRunnable = null
    }

    private fun sendTimeUpdate() {
        val player = player ?: return

        val currentTimeMs = player.currentPosition
        val durationMs = player.duration.takeIf { it > 0 } ?: 0L
        val bufferedMs = player.bufferedPosition

        // Handle A-B loop
        if (loopEnabled && currentTimeMs >= loopEndMs) {
            seekToInternal(loopStartMs)
            return
        }

        sendEvent("onTimeUpdate", Arguments.createMap().apply {
            putDouble("currentTimeMs", currentTimeMs.toDouble())
            putDouble("durationMs", durationMs.toDouble())
            putDouble("bufferedMs", bufferedMs.toDouble())
        })
    }

    // MARK: - Playback Control

    @ReactMethod
    fun play() {
        handler?.post {
            player?.play()
        }
    }

    @ReactMethod
    fun pause() {
        handler?.post {
            player?.pause()
        }
    }

    @ReactMethod
    fun stop() {
        handler?.post {
            cleanup()
            updateStatus("idle")
        }
    }

    @ReactMethod
    fun seekTo(timeMs: Double) {
        handler?.post {
            seekToInternal(timeMs.toLong())
        }
    }

    private fun seekToInternal(timeMs: Long) {
        player?.seekTo(timeMs)
    }

    @ReactMethod
    fun setRate(rate: Double) {
        handler?.post {
            player?.setPlaybackSpeed(rate.toFloat())
        }
    }

    @ReactMethod
    fun setLoop(startMs: Double, endMs: Double, enabled: Boolean) {
        loopStartMs = startMs.toLong()
        loopEndMs = endMs.toLong()
        loopEnabled = enabled
    }

    @ReactMethod
    fun setVolume(volume: Double) {
        handler?.post {
            player?.volume = volume.toFloat().coerceIn(0f, 1f)
        }
    }

    // MARK: - Synchronous Getters

    @ReactMethod(isBlockingSynchronousMethod = true)
    fun getCurrentTimeMs(): Double {
        return player?.currentPosition?.toDouble() ?: 0.0
    }

    @ReactMethod(isBlockingSynchronousMethod = true)
    fun getDurationMs(): Double {
        val duration = player?.duration ?: 0L
        return if (duration > 0) duration.toDouble() else 0.0
    }

    @ReactMethod(isBlockingSynchronousMethod = true)
    fun getStatus(): String {
        return currentStatus
    }

    // MARK: - Event Emission

    private fun sendEvent(eventName: String, params: WritableMap) {
        reactContext
            .getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter::class.java)
            .emit(eventName, params)
    }

    private fun updateStatus(status: String) {
        currentStatus = status
        sendEvent("onStatusChange", Arguments.createMap().apply {
            putString("status", status)
        })
    }

    private fun sendError(code: String, message: String) {
        currentStatus = "error"
        sendEvent("onStatusChange", Arguments.createMap().apply {
            putString("status", "error")
            putString("error", message)
        })
        sendEvent("onError", Arguments.createMap().apply {
            putString("code", code)
            putString("message", message)
        })
    }

    // MARK: - Lifecycle

    override fun onHostResume() {}

    override fun onHostPause() {}

    override fun onHostDestroy() {
        cleanup()
    }

    private fun cleanup() {
        stopTimeUpdates()
        player?.removeListener(playerListener)
        player?.release()
        player = null
        loopEnabled = false
        loopStartMs = 0
        loopEndMs = 0
    }

    // Required for event emitter
    @ReactMethod
    fun addListener(eventName: String) {}

    @ReactMethod
    fun removeListeners(count: Int) {}
}

File: apps/mobile/ios/Audixa/NativePlayerModule.swift
import Foundation
import AVFoundation
import React

/**
 * Native Player Module for iOS
 * 
 * Provides high-performance media playback using AVPlayer with:
 * - Precise seeking (ms-level accuracy)
 * - A-B loop support
 * - Background audio playback
 * - Lock screen controls
 */
@objc(NativePlayerModule)
class NativePlayerModule: RCTEventEmitter {
  
  // MARK: - Properties
  
  private var player: AVPlayer?
  private var playerItem: AVPlayerItem?
  private var timeObserver: Any?
  private var statusObserver: NSKeyValueObservation?
  
  private var loopEnabled = false
  private var loopStartMs: Double = 0
  private var loopEndMs: Double = 0
  
  private var currentStatus: String = "idle"
  
  // Time update frequency (50ms = 20Hz)
  private let timeUpdateInterval = CMTime(value: 1, timescale: 20)
  
  // MARK: - RCTEventEmitter
  
  override static func moduleName() -> String! {
    return "NativePlayerModule"
  }
  
  override func supportedEvents() -> [String]! {
    return ["onTimeUpdate", "onStatusChange", "onBuffering", "onEnded", "onError"]
  }
  
  override static func requiresMainQueueSetup() -> Bool {
    return true
  }
  
  // MARK: - Audio Session Configuration
  
  private func configureAudioSession() {
    do {
      let session = AVAudioSession.sharedInstance()
      try session.setCategory(.playback, mode: .default, options: [])
      try session.setActive(true)
    } catch {
      print("Failed to configure audio session: \(error)")
    }
  }
  
  // MARK: - Player Management
  
  @objc func load(_ uri: String, autoPlay: Bool) {
    DispatchQueue.main.async { [weak self] in
      self?.loadInternal(uri: uri, autoPlay: autoPlay)
    }
  }
  
  private func loadInternal(uri: String, autoPlay: Bool) {
    // Clean up previous player
    cleanup()
    
    // Configure audio session for background playback
    configureAudioSession()
    
    // Create URL
    guard let url = URL(string: uri) ?? URL(fileURLWithPath: uri) as URL? else {
      sendError(code: "INVALID_URI", message: "Invalid media URI")
      return
    }
    
    // Create player item
    playerItem = AVPlayerItem(url: url)
    player = AVPlayer(playerItem: playerItem)
    
    // Set up observers
    setupObservers()
    
    // Update status
    updateStatus("loading")
    
    // Auto-play if requested
    if autoPlay {
      player?.play()
    }
  }
  
  private func setupObservers() {
    guard let player = player, let playerItem = playerItem else { return }
    
    // Time observer
    timeObserver = player.addPeriodicTimeObserver(
      forInterval: timeUpdateInterval,
      queue: .main
    ) { [weak self] time in
      self?.handleTimeUpdate(time: time)
    }
    
    // Status observer
    statusObserver = playerItem.observe(\.status, options: [.new]) { [weak self] item, _ in
      DispatchQueue.main.async {
        self?.handleStatusChange(status: item.status)
      }
    }
    
    // End of playback notification
    NotificationCenter.default.addObserver(
      self,
      selector: #selector(playerDidFinishPlaying),
      name: .AVPlayerItemDidPlayToEndTime,
      object: playerItem
    )
  }
  
  private func handleTimeUpdate(time: CMTime) {
    guard let playerItem = playerItem else { return }
    
    let currentTimeMs = time.seconds * 1000
    let durationMs = playerItem.duration.seconds * 1000
    let bufferedMs = getBufferedTime() * 1000
    
    // Handle A-B loop
    if loopEnabled && currentTimeMs >= loopEndMs {
      seekToInternal(timeMs: loopStartMs)
      return
    }
    
    sendEvent(withName: "onTimeUpdate", body: [
      "currentTimeMs": currentTimeMs,
      "durationMs": durationMs.isNaN ? 0 : durationMs,
      "bufferedMs": bufferedMs
    ])
  }
  
  private func handleStatusChange(status: AVPlayerItem.Status) {
    switch status {
    case .readyToPlay:
      updateStatus("paused")
    case .failed:
      let message = playerItem?.error?.localizedDescription ?? "Playback failed"
      sendError(code: "PLAYBACK_ERROR", message: message)
    case .unknown:
      break
    @unknown default:
      break
    }
  }
  
  @objc private func playerDidFinishPlaying() {
    if loopEnabled {
      seekToInternal(timeMs: loopStartMs)
      player?.play()
    } else {
      updateStatus("ended")
      sendEvent(withName: "onEnded", body: [:])
    }
  }
  
  // MARK: - Playback Control
  
  @objc func play() {
    DispatchQueue.main.async { [weak self] in
      self?.player?.play()
      self?.updateStatus("playing")
    }
  }
  
  @objc func pause() {
    DispatchQueue.main.async { [weak self] in
      self?.player?.pause()
      self?.updateStatus("paused")
    }
  }
  
  @objc func stop() {
    DispatchQueue.main.async { [weak self] in
      self?.cleanup()
      self?.updateStatus("idle")
    }
  }
  
  @objc func seekTo(_ timeMs: Double) {
    DispatchQueue.main.async { [weak self] in
      self?.seekToInternal(timeMs: timeMs)
    }
  }
  
  private func seekToInternal(timeMs: Double) {
    let time = CMTime(seconds: timeMs / 1000, preferredTimescale: 1000)
    
    // Use precise seeking with zero tolerance
    player?.seek(
      to: time,
      toleranceBefore: .zero,
      toleranceAfter: .zero
    )
  }
  
  @objc func setRate(_ rate: Double) {
    DispatchQueue.main.async { [weak self] in
      self?.player?.rate = Float(rate)
    }
  }
  
  @objc func setLoop(_ startMs: Double, endMs: Double, enabled: Bool) {
    loopStartMs = startMs
    loopEndMs = endMs
    loopEnabled = enabled
  }
  
  @objc func setVolume(_ volume: Double) {
    DispatchQueue.main.async { [weak self] in
      self?.player?.volume = Float(max(0, min(1, volume)))
    }
  }
  
  // MARK: - Synchronous Getters
  
  @objc func getCurrentTimeMs() -> Double {
    guard let player = player else { return 0 }
    return player.currentTime().seconds * 1000
  }
  
  @objc func getDurationMs() -> Double {
    guard let playerItem = playerItem else { return 0 }
    let duration = playerItem.duration.seconds
    return duration.isNaN ? 0 : duration * 1000
  }
  
  @objc func getStatus() -> String {
    return currentStatus
  }
  
  // MARK: - Helpers
  
  private func getBufferedTime() -> Double {
    guard let timeRange = playerItem?.loadedTimeRanges.first?.timeRangeValue else {
      return 0
    }
    return CMTimeGetSeconds(CMTimeAdd(timeRange.start, timeRange.duration))
  }
  
  private func updateStatus(_ status: String) {
    currentStatus = status
    sendEvent(withName: "onStatusChange", body: [
      "status": status
    ])
  }
  
  private func sendError(code: String, message: String) {
    currentStatus = "error"
    sendEvent(withName: "onStatusChange", body: [
      "status": "error",
      "error": message
    ])
    sendEvent(withName: "onError", body: [
      "code": code,
      "message": message
    ])
  }
  
  private func cleanup() {
    if let observer = timeObserver {
      player?.removeTimeObserver(observer)
      timeObserver = nil
    }
    
    statusObserver?.invalidate()
    statusObserver = nil
    
    NotificationCenter.default.removeObserver(self)
    
    player?.pause()
    player = nil
    playerItem = nil
    
    loopEnabled = false
    loopStartMs = 0
    loopEndMs = 0
  }
  
  deinit {
    cleanup()
  }
}
